Port MemoryMappedFileSecurity and add extensions for MemoryMappedFile

carlossanlop opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/941

We did not port `System.IO.MemoryMappedFiles.MemoryMappedFileSecurity` to Core as it wasn't cross-plat. There is still a need to set security on memory mapped files for Windows, just as there is for files/directories/etc. We should port these to `System.IO.FileSystem.AccessControl` and add extension methods for `MemoryMappedFile` there so Windows specific codebases can be ported and retain their security model.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/Security/AccessControl/MemoryMappedFileSecurity.cs:
```diff
@@ -0,0 +1,119 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32.SafeHandles;
```
This file is WIP.

carlossanlop replied:
I filled this file, although I'm unsure if I need to add more method overrides. I'll continue comparing this with the existing derived classes we already have.

carlossanlop replied:
Update: After deeper investigation, this should be all I need to include in this file.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -515,5 +390,94 @@ private static void CleanupFile(FileStream fileStream, bool existed, string path
             }
         }
+
+        private static void ValidateParametersCreateFromFile(FileStream fileStream, string? mapName, long capacity,
```
I created 3 different parameter validation methods because they are not exactly equal, and I didn't want to introduce breaking changes by either changing the type of exception thrown for certain cases (like the mapName null check), or by changing the order. Thoughts?

carlossanlop replied:
Not necessary anymore.

jkotas commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,101 @@
+        {
+            fixed (byte* pSecurityDescriptor = memoryMappedFileSecurity.GetSecurityDescriptorBinaryForm())
+            {
+                return GetMemoryMappedFileWithTypeMemberInvocation("CreateFromFile",
```
Can we add public APIs that take a handle instead of using this reflection hack?

JeremyKuhne replied:
We can look into making the [private constructors](https://source.dot.net/#System.IO.MemoryMappedFiles/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,28) public. If we do, we would want to include the core logic in both assemblies to avoid code duplication (to reduce the risk of mismatching with the other APIs).

carlossanlop replied:
What about adding a setter to the [public SafeMemoryMappedFileHandle property](https://source.dot.net/#System.IO.MemoryMappedFiles/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,482)?

The reason why I suggest this alternative is because the ACL APIs we added for Mutex, Semaphore and EventWaitHandle all have that setter, and none have a public constructor that receives a handle.

Since we had the public setter, we were able to:
1. Create the handle with the appropriate security.
2. Create the object.
3. Switch its inner handle with the one we created.

[Example for EventWaitHandle](https://source.dot.net/#System.Threading.AccessControl/System/Threading/EventWaitHandleAcl.cs,64).

jkotas replied:
> Mutex, Semaphore and EventWaitHandle

WaitHandle that all these inherit from is a thin wrapper around the underlying OS handle. WaitHandle has just the safe handle field and nothing else. It makes it reasonable to have a setter for the safe handle.

`MemoryMappedFile` is not a thin wrapper like that. It has multiple fields that better stay in sync. It would be hard to reason about the behavior of the public setter given this.

jkotas commented on src/libraries/System.IO.FileSystem.AccessControl/src/System.IO.FileSystem.AccessControl.csproj:
```diff
@@ -47,14 +47,16 @@
-    <Compile Include="$(CommonPath)System\IO\FileSystem.Attributes.Windows.cs" Link="Common\System\IO\FileSystem.Attributes.Windows.cs" />
-    <Compile Include="$(CommonPath)System\IO\FileSystem.DirectoryCreation.Windows.cs" Link="Common\System\IO\FileSystem.DirectoryCreation.Windows.cs" />
+    <Compile Include="System\IO\FileSystemAclExtensions.cs" />
+    <Compile Include="System\IO\MemoryMappedFiles\MemoryMappedFileAcl.cs" />
```
This section is used for netstandard2.0 too, but the reflection hack in MemoryMappedFileAcl.cs is only going to work on .NET 5+.

carlossanlop replied:
Pending to address this.

carlossanlop replied:
Since I'm not using a reflection hack anymore, I think these files will not have a compatibility problem, but I'll double check with Eric.

jkotas commented on src/libraries/System.Private.CoreLib/src/Interop/Windows/Kernel32/Interop.SECURITY_ATTRIBUTES.cs:
```diff
@@ -16,5 +16,24 @@ internal struct SECURITY_ATTRIBUTES
+            };
+        }
+
+        internal static unsafe SECURITY_ATTRIBUTES GetSecAttrs(byte* pSecurityDescriptor, BOOL inheritHandle = default)
```
If you are making a helper method out of this, the argument should be `bool` so that callers do not have to normalize it.

jkotas replied:
It may look better to make these to be `SECURITY_ATTRIBUTES` constructors.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/ref/System.IO.FileSystem.AccessControl.cs:
```diff
@@ -118,4 +120,37 @@ public abstract partial class FileSystemSecurity : System.Security.AccessControl
         public void SetAccessRule(System.Security.AccessControl.FileSystemAccessRule rule) { }
         public void SetAuditRule(System.Security.AccessControl.FileSystemAuditRule rule) { }
     }
+    public class MemoryMappedFileSecurity : System.Security.AccessControl.NativeObjectSecurity
```
@JeremyKuhne I am going to have to add a couple of new classes: `MemoryMappedFileAccessRule` and `MemoryMappedFileAuditRule`. They did not use to exist in .NET Framework. Should we submit them for API review?

carlossanlop replied:
After deeper investigation, I concluded that they will not be necessary.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.cs:
```diff
@@ -14,23 +14,62 @@ public partial class MemoryMappedFile : IDisposable
+        /// <param name="leaveOpen">Whether the file stream should be left open or not.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="handle" /> or <paramref name="fileStream" /> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentException"><paramref name="handle" /> is closed or invalid.</exception>
+        public MemoryMappedFile(SafeMemoryMappedFileHandle handle, FileStream fileStream, bool leaveOpen)
```
Update: Before I publish this draft PR, I will have to bring this other private constructor to API Approval to make it public. 

This constructor, along with the other one that receives only the handle (and has already been approved to be made public), are needed internally when creating MemoryMappedFiles via CreateOrOpen, CreateNew, CreateFromFile.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,434 @@
+            fixed (byte* pSecurityDescriptor = memoryMappedFileSecurity.GetSecurityDescriptorBinaryForm())
+            {
+                SafeMemoryMappedFileHandle handle = CreateCore(fileStream, mapName, access, MemoryMappedFileOptions.None, capacity, GetSecAttrs(pSecurityDescriptor, inheritability));
+                return new MemoryMappedFile(handle, fileStream, leaveOpen);
```
This is the second constructor that needs to be approved to be made public.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,434 @@
+            fixed (byte* pSecurityDescriptor = memoryMappedFileSecurity.GetSecurityDescriptorBinaryForm())
+            {
+                SafeMemoryMappedFileHandle handle = CreateCore(null, mapName, access, options, capacity, GetSecAttrs(pSecurityDescriptor, inheritability));
+                return new MemoryMappedFile(handle);
```
This is the constructor that was already approved to be made public.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,434 @@
+                }
+            }
+
+            return new MemoryMappedFile(handle);
```
This is the constructor that was already approved to be made public.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,434 @@
+
+namespace System.IO.MemoryMappedFiles
+{
+    public static class MemoryMappedFileAcl
```
The `CreateFromFile`, `CreateNew` and `CreateOrOpen` methods are almost identical to the ones that already exist in the `MemoryMappedFile` class. They only differ in the section that creates the file handle with the provided ACL configuration, which uses the two constructors that had to be made public.

carlossanlop replied:
To avoid requesting to make the `MemoryMappedFile` constructor that receives 3 arguments public, I am calling the one that was approved, and indicating in the documentation that the `leaveOpen` argument is unused: the handle will always be left open after the memory-mapped file is disposed.

carlossanlop replied:
Update: I asked @bartonjs for his advice on this, and the other constructor can be approved if the second argument is not a file stream, but an `IDisposable` (since the only thing we need from that object is its `Dispose()` method).
The code has been updated to reflect this, and will bring the API to review on the next meeting.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,434 @@
+
+namespace System.IO.MemoryMappedFiles
+{
+    public static class MemoryMappedFileAcl
```
Pending adding unit tests.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,434 @@
+    public static class MemoryMappedFileAcl
+    {
+        private const int DefaultSize = 0;
+        private static Type MemoryMappedFileType { get; } = typeof(MemoryMappedFile);
```
Not needed anymore. Need to remove.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/ref/System.IO.FileSystem.AccessControl.cs:
```diff
@@ -118,4 +118,21 @@ public abstract partial class FileSystemSecurity : System.Security.AccessControl
         public void SetAccessRule(System.Security.AccessControl.FileSystemAccessRule rule) { }
         public void SetAuditRule(System.Security.AccessControl.FileSystemAuditRule rule) { }
     }
+    public class MemoryMappedFileSecurity : ObjectSecurity<System.IO.MemoryMappedFiles.MemoryMappedFileRights>
```
I noticed I placed this class in the wrong namespace. I have to move it to System.IO.MemoryMappedFiles, like we have it in [netfx](https://apisof.net/catalog/System.IO.MemoryMappedFiles.MemoryMappedFileSecurity).

carlossanlop commented on src/libraries/Common/src/System/IO/MemoryMappedFileInternal.Windows.cs:
```diff
@@ -2,43 +2,50 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using Microsoft.Win32.SafeHandles;
```
MemoryMappedFile.Windows.cs should stay, and that file's methods should call the MemoryMappedFileInternal ones, otherwise the Unix build fails.

carlossanlop committed:
Port MemoryMappedFileSecurity and add extensions for MemoryMappedFile

jkotas commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.CreateFileMapping.cs:
```diff
@@ -28,5 +57,12 @@ internal partial class Kernel32
+        public static void SplitLong(long number, out int high, out int low)
+        {
+            high = unchecked((int)(number >> 32));
+            low = unchecked((int)(number & 0x00000000FFFFFFFFL));
```
```suggestion
            low = unchecked((int)(number));
```

jkotas commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/Interop.Windows.cs:
```diff
@@ -33,9 +32,9 @@ public static unsafe void CheckForAvailableVirtualMemory(ulong nativeSize)
         // split the long into two ints
         int capacityHigh, capacityLow;
-        SplitLong(maximumSize, out capacityHigh, out capacityLow);
+        Kernel32.SplitLong(maximumSize, out capacityHigh, out capacityLow);
```
Can these use the Kernel32 wrapper that takes the `long` instead?

jkotas commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.CreateFileMapping.cs:
```diff
@@ -5,12 +5,41 @@
+            int capacityHigh, capacityLow;
+            SplitLong(maximumSize, out capacityHigh, out capacityLow);
+
+            return CreateFileMapping(hFile, ref securityAttributes, pageProtection, capacityHigh, capacityLow, name);
```
```suggestion
            return CreateFileMapping(hFile, ref securityAttributes, pageProtection, dwMaximumSizeHigh: (int)(maximumSize >> 32), dwMaximumSizeLow: (int)maximumSize , name);
```

Now that the split is in the wrapper, I think it is better to do it inline. The extra ceremony from the SplitLong method does seems worth it. It is just a pile of busy work for the JIT to optimize out.

jkotas commented on src/libraries/System.IO.MemoryMappedFiles/src/System.IO.MemoryMappedFiles.csproj:
```diff
@@ -7,157 +7,79 @@
-    <Compile Include="$(CommonPath)System\IO\Win32Marshal.cs">
-      <Link>Common\System\IO\Win32Marshal.cs</Link>
-    </Compile>
+    <Compile Include="$(CommonPath)Interop\Windows\Interop.BOOL.cs" Link="Common\Interop\Windows\Interop.BOOL.cs" />
```
IMHO, this reformatting is questionable. It produces super long lines that are hard to read - see discussion in https://github.com/dotnet/runtime/issues/1266

carlossanlop replied:
The reason why I did this is because it's easier to sort single lines alphabetically. Once the `<Link></Link>` is introduced, it gets formatted in a new line, making it difficult to sort.

But if the idea is to unify and avoid longer lines, I'll revert this change.

stephentoub commented:
@carlossanlop, why is this marked as a draft?

carlossanlop commented:
> @carlossanlop, why is this marked as a draft?

@stephentoub it wasn't ready before my last commit. I just forgot to mark it as ready for review. It is (mostly) ready now. Thanks for noticing.

stephentoub commented:
> it wasn't ready before my last commit. I just forgot to mark it as ready for review.

Ok, thanks. Why are we opening PRs at all before they're ready for review?

carlossanlop commented:
> Why are we opening PRs at all before they're ready for review?

@stephentoub I was mostly missing unit tests and I had yet to do some code cleaning.

I opened it as a draft PR (which Github considers the opposite of "ready for review"). I noticed I get some helpful comments while I am working on my changes (particularly from Jan), and I think it also helps confirm that someone is already working on a particular issue: work would not get duplicated, and interested parties can see the process and comment on them. 

But if we don't want to add drafts to the queue, that's fine, just let me know. I guess I can still ask for feedback if I just share a commit from my fork to whoever is involved in my changes.

carlossanlop commented on src/libraries/Common/src/System/IO/MemoryMappedFileInternal.Windows.cs:
```diff
@@ -0,0 +1,280 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
This file used to be the Windows partial class for `MemoryMappedFile`, but since a big chunk of the code in that file needed to be used by the new methods, I decided to move the shared code to a new Common class called `MemoryMappedFileInternal` to avoid code duplication.

carlossanlop commented on src/libraries/Common/src/System/IO/MemoryMappedFileInternal.cs:
```diff
@@ -0,0 +1,167 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
This file also contains code that is shared between the new methods and `MemoryMappedFile`. The code is also used in Unix, which is why the class had to be split as a partial.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs:
```diff
@@ -0,0 +1,205 @@
+            fixed (byte* pSecurityDescriptor = memoryMappedFileSecurity.GetSecurityDescriptorBinaryForm())
+            {
+                SafeMemoryMappedFileHandle handle = MemoryMappedFileInternal.CreateCore(fileStream, mapName, access, MemoryMappedFileOptions.None, updatedCapacity, MemoryMappedFileInternal.GetSecAttrs(pSecurityDescriptor, inheritability));
+                return new MemoryMappedFile(handle); // This constructor leaves the handle open
```
The user will have to take care of manually disposing the passed `FileStream` after this `MemoryMappedFile` object is disposed.

carlossanlop replied:
Note: The `MemoryMappedFile(SafeMemoryMappedFileHandle, FileStream, Boolean)` is not being used right now because it wasn't approved in the API Proposal meeting. It's what we currently [use in Framework](https://referencesource.microsoft.com/#System.Core/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,244), and what we use in the [existing .NET Core method](https://source.dot.net/#System.IO.MemoryMappedFiles/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,254).

After discussing this with @JeremyKuhne and @ericstj, there seem to be enough reasons to go to the API Proposal meeting again to request making this constructor public - It may be an ugly pattern, but we have precedent of other classes using it, and we don't have a better choice 
 - we should not use reflection to call the private constructor, and we should not leave a fileStream open even if the user indicated they want it closed.

Other examples of public constructors that use a similar pattern:

- https://source.dot.net/#System.Private.CoreLib/BinaryReader.cs,49
- https://source.dot.net/#System.Private.CoreLib/BinaryWriter.cs,51

cc @bartonjs @stephentoub @GrabYourPitchforks

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileSecurity.cs:
```diff
@@ -0,0 +1,28 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
This file was brought from Framework.

danmosemsft commented on src/libraries/Common/src/System/IO/MemoryMappedFileInternal.cs:
```diff
@@ -0,0 +1,167 @@
+
+            if (IntPtr.Size == 4 && capacity > uint.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_CapacityLargerThanLogicalAddressSpaceNotAllowed);
```
Aside - feel free to ignore here. In general, it is nice for strings describing invalid arguments to include the value itself. I have found this can help to diagnose issues when you are presented with only the exception message (such as from a server log).

Among the strings like this are:
```
  <data name="ArgumentOutOfRange_CapacityLargerThanLogicalAddressSpaceNotAllowed" xml:space="preserve">
    <value>The capacity cannot be greater than the size of the system's logical address space.</value>
  </data>
  <data name="ArgumentOutOfRange_FileLengthTooBig" xml:space="preserve">
    <value>Specified file length was too large for the file system.</value>
  </data>
  <data name="ArgumentOutOfRange_NeedPositiveNumber" xml:space="preserve">
    <value>A positive number is required.</value>
  </data>
  <data name="ArgumentOutOfRange_PositiveOrDefaultCapacityRequired" xml:space="preserve">
    <value>The capacity must be greater than or equal to 0. 0 represents the size of the file being mapped.</value>
  </data>
  <data name="ArgumentOutOfRange_PositiveOrDefaultSizeRequired" xml:space="preserve">
    <value>The size must be greater than or equal to 0. If 0 is specified, the view extends from the specified offset to the end of the file mapping.</value>
  </data>
  <data name="ArgumentOutOfRange_CapacityGEFileSizeRequired" xml:space="preserve">
    <value>The capacity may not be smaller than the file size.</value>
  </data>
```

This wasn't generally historical practice in the BCL so there are many of these. But going forward (again not necessarily in this PR) it would be nice to improve as we go along.

danmosemsft replied:
Note that one way of doing this is to use the overloads that take a value - they append the bad value to the message you supply. I personally prefer the value in the string as sometimes the two lines of the message get separated in the log, but that's not a big deal.

carlossanlop replied:
Just noticed this comment. Thanks for your suggestion. I will address it on the next iteration.

carlossanlop commented on src/libraries/System.IO.FileSystem.AccessControl/tests/MemoryMappedFileAclTests.cs:
```diff
@@ -0,0 +1,532 @@
+            Assert.NotNull(mmf);
+        }
+
+        //[Theory]
```
I'm investigating why this unit test is not passing. The security object extracted via `GetAccessControl` is not the same as the one I passed to the `Create*` methods.

stephentoub commented:
Thanks, Carlos.

> I was mostly missing unit tests and I had yet to do some code cleaning.

Ok.  I think it'd be better in the future if we not open PRs when there are tests known to be missing.  That makes it all the more likely that there are bugs in what's being PR'd, since there are known test holes that need to be filled.

> I opened it as a draft PR (which Github considers the opposite of "ready for review").

Right.  As the opposite of "ready for review", there are relatively few times when a draft PR makes sense, IMHO.  If the goal is to get feedback on the general approach, ok, but having such a PR open for almost three months is problematic.  When there's just one of them, it's not a huge deal, but we've got 8 pages of PRs right now, which makes it really easy to stop paying attention to most of them.

> I think it also helps confirm that someone is already working on a particular issue: work would not get duplicated, and interested parties can see the process and comment on them.

I think that's what issues and issue assignment is for.

carlossanlop commented:
Understood, no problem @stephentoub. Thanks for the explanation.
I'll close the issue while it's in draft mode. I'll reopen when the unit tests are ready.

stephentoub commented:
Thanks, appreciate it :)

carlossanlop committed:
Address PR comments, add code to test potentially new constructor that takes an IDisposable, add unit tests to ensure security objects are verified after MMF creation.

Fix minor unit test failures

Remove unnecessary leaveOpen in new MemoryMappedFile constructor. Add unit test that disposes FileStream after creating the MemoryMappedFile, but before reading the MemoryMappedFile contents.

Fix unit tests that fail or not fail when security is null.

carlossanlop commented:
@JeremyKuhne @bartonjs

Reopening. The tests are passing both in NetCoreApp and NetFX (locally).

I added the code necessary to get this PR working, but it will still need an extra API Proposal, which I will submit in a separate issue:

```cs
namespace System.IO.MemoryMappedFiles
{
    public class MemoryMappedFile : IDisposable
    {
        public MemoryMappedFile(SafeMemoryMappedFileHandle handle, [NullableAttribute(2)] IDisposable? fileStream, bool leaveOpen);
    }
}
```

This public constructor is needed to allow `MemoryMappedFileAcl.CreateFromFile` to create a `MemoryMappedFile`.

The constructor needs to pass the `handle` of the file stream created by the user, the `fileStream` itself (casted to an `IDisposable`) and a boolean that indicates if the user wants to keep the `fileStream` open or not, after the `MemoryMappedFile` is disposed.

The `fileStream` needs to be passed to ensure it gets rooted. It cannot be passed as `null`.
The `leaveOpen` boolean is what indicates if the passed disposable object should be disposed or not.

carlossanlop commented on src/libraries/System.IO.MemoryMappedFiles/ref/System.IO.MemoryMappedFiles.cs:
```diff
@@ -24,6 +24,8 @@ namespace System.IO.MemoryMappedFiles
     {
         internal MemoryMappedFile() { }
+        public MemoryMappedFile(Microsoft.Win32.SafeHandles.SafeMemoryMappedFileHandle handle) { }
+        public MemoryMappedFile(Microsoft.Win32.SafeHandles.SafeMemoryMappedFileHandle handle, System.IDisposable? fileStream, bool leaveOpen) { }
```
Need to bring this new constructor to API review.

bartonjs replied:
I don't know that `leaveOpen` is useful, or needed... Generally it's used because the disposable parameter is being used as its strong type (e.g. data being read from a Stream).  For this constructor, it seems like if you would pass false, just pass null for the parent (currently named fileStream, but that's not a great name, since it can be anything)

carlossanlop replied:
I shared your initial suggestion to @JeremyKuhne about passing `null` and removing `leaveOpen`.

The reason why we kept the boolean is because we need to root the file stream to the `MemoryMappedFile`. We can't pass `null`. The boolean will tell the constructor if the passed disposable object should be disposed or not.

When the user decides the object should be left open, then we pass a disposable wrapper class that will root the fileStream inside it.

bartonjs replied:
> When the user decides the object should be left open, then we pass a disposable wrapper class that will root the fileStream inside it.

Right, so you don't need the `leaveOpen` parameter on the constructor :smile:.  You can stay rooted by passing an IDisposable wrapper that just doesn't forward the call to Dispose.

carlossanlop replied:
🤓 Gah, you're right. I'll fix that.

JeremyKuhne replied:
While the wrapper workaround works, I'm concerned about having this take an `IDisposable`. Users would be forced to create the wrapper as well when they want to not have the `MemoryMappedFile` close the backing stream.  Documentation would be something like this:

| Disposable for the the `FileStream` used to open the `SafeMemoryMappedFileHandle`, if any. The disposable will be disposed when the `MemoryMappedFile` is disposed.
| Notes: If you need to keep the source `FileStream` (if any) open after the `MemoryMappedFile` is disposed, either don't pass the handle in or wrap it in an `IDisposable` wrapper class that doesn't call dispose.

@carlossanlop Can you test what happens when using the `MemoryMappedFile` after the opening `FileStream` is disposed? Given the old API surface my presumption is that the `FileStream` _has_ to be left open. I haven't validated that, however. _If_ `MemoryMappedFile` works fine with the `FileStream` closed, then I'm much less concerned about not being more generic in the API. If it breaks using the `MemoryMappedFile` then I'd be even more inclined to use `FileStream, bool` like we do everywhere else. If that can't get approved we'll need to be very explicit in the docs about rooting the `FileStream` to avoid finalizer issues.

carlossanlop replied:
> _If MemoryMappedFile works fine with the FileStream closed, then I'm much less concerned about not being more generic in the API._

I added a new unit test that:

1. Creates a `FileStream`.
2. Creates a new `MemoryMappedFileStream` using the new `CreateFromFile` method, by consuming that `FileStream`.
3. Disposes the `DileStream`.
4. Creates a `MemoryMappedViewAccessor` and reads some bytes from it, then disposes it.
5. Creates a `MemoryMappedViewStream` and reads some bytes from it, then disposes it.
5. Disposes the `MemoryMappedViewAccessor`.

The unit test passed.

carlossanlop commented:
The CI hit a bunch of unrelated Linux failures coming from System.Runtime.Serialization.Formatters. This change is Windows only.

carlossanlop committed:
Try suggestion #1 from approved API proposal: Move implementation to MemoryMappedFiles. Expose the refs in FileSystem.AccessControl.

carlossanlop commented:
Update on the latest change: I applied suggestion #1 in the approved API proposal [comment](https://github.com/dotnet/runtime/issues/33206#issuecomment-610526663):

> _Option 1: Move the implementation from the higher level assembly to the assembly containing MemoryMappedFile (while the ref assembly setup)_

Kudos to @safern for helping me figuring out how to get the assemblies to work with each other. cc @ericstj 

I built all configurations and the build succeeded. Now waiting for CI.

jkotas commented on src/libraries/System.IO.FileSystem.AccessControl/tests/AclTestBase.cs:
```diff
@@ -0,0 +1,62 @@
+﻿using System.Collections.Generic;
```
Missing license header

carlossanlop replied:
Fixed

jkotas commented on src/libraries/System.IO.MemoryMappedFiles/ref/System.IO.MemoryMappedFiles.csproj:
```diff
@@ -12,5 +12,6 @@
     <ProjectReference Include="..\..\System.IO.UnmanagedMemoryStream\ref\System.IO.UnmanagedMemoryStream.csproj" />
     <ProjectReference Include="..\..\System.Runtime.InteropServices\ref\System.Runtime.InteropServices.csproj" />
     <ProjectReference Include="..\..\System.Runtime\ref\System.Runtime.csproj" />
+    <ProjectReference Include="..\..\System.Security.AccessControl\ref\System.Security.AccessControl.csproj" />
```
So now you have `System.IO.MemoryMappedFiles` (part of the netcoreapp refs, does not ship as package) depending on `System.Security.AccessControl` (still ships as a package).  That does not look right.

E.g. I believe that this means that `System.Security.AccessControl` ref needs to be part of what you get by default for netcoreapp. But `System.Security.AccessControl` is Windows-specific and the whole point of this exercise was to not fault it into the platform-agnostic netcoreapp by default.

safern replied:
I don't think this is needed anymore. This seems like a leftover from when we started figuring out the structure of this.

carlossanlop replied:
Removed. I built locally and it succeeded.

carlossanlop committed:
Add missing comment header, remove unnecessary dependency from csproj.

jkotas commented on src/libraries/System.IO.FileSystem.AccessControl/ref/System.IO.FileSystem.AccessControl.cs:
```diff
@@ -58,18 +77,18 @@ public sealed partial class FileSecurity : System.Security.AccessControl.FileSys
     }
     public sealed partial class FileSystemAccessRule : System.Security.AccessControl.AccessRule
     {
-        public FileSystemAccessRule(System.Security.Principal.IdentityReference identity, System.Security.AccessControl.FileSystemRights fileSystemRights, System.Security.AccessControl.AccessControlType type) : base (default(System.Security.Principal.IdentityReference), default(int), default(bool), default(System.Security.AccessControl.InheritanceFlags), default(System.Security.AccessControl.Prop ...
```
Are these formatting changes intentional? IIRC, `/t:GenerateReferenceSource` generates extra space after `base`.

carlossanlop replied:
I did not execute a command using `/t:GenerateReferenceSource`. What's the full command I should run?
I probably just need to revert the change.

safern replied:
After building the src and having the src shape that you want you need to run `dotnet msbuild /t:GenerateReferenceSource` in the source project.

jkotas commented on src/libraries/System.IO.MemoryMappedFiles/src/System.IO.MemoryMappedFiles.csproj:
```diff
@@ -132,32 +139,31 @@
     <Reference Include="System.Runtime.InteropServices" />
-    <Reference Include="System.Threading" />
+    <Reference Include="System.Runtime" />
+    <Reference Include="System.Security.AccessControl" />
```
This is again dependency from inbox binary to a binary shipping as package. Are all combinations of this going to work well?

safern replied:
I think this is fine. `System.Security.AccessControl` is OOB only for the ref assembly but the implementation is part of the shared framework:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.AccessControl/Directory.Build.props#L5-L6

Given that we have some inbox projects already depending on it:
https://github.com/dotnet/runtime/blob/02ea20ae6dd0450d04215b561eae449ae1ff289e/src/libraries/System.IO.Pipes/src/System.IO.Pipes.csproj#L125

If I remember correctly we made AccessControl that way specifically to be able to reference it from inbox binaries.
cc: @ericstj

jkotas commented on src/libraries/System.IO.MemoryMappedFiles/ref/System.IO.MemoryMappedFiles.cs:
```diff
@@ -24,6 +24,8 @@ namespace System.IO.MemoryMappedFiles
     public partial class MemoryMappedFile : System.IDisposable
     {
         internal MemoryMappedFile() { }
+        internal MemoryMappedFile(Microsoft.Win32.SafeHandles.SafeMemoryMappedFileHandle handle) { }
```
Do we really need internal APIs in ref file?

bartonjs replied:
The zero argument one, yeah, since it makes the type not appear to have a default public ctor.  The other two I agree shouldn't be here.

Is this file being hand-edited, vs generated with `dotnet msbuild /t:GenerateReferenceSource` from the src directory (after building the src directory)?

carlossanlop replied:
It was hand edited. I did not know I had to run that command. I'll do that now.

jkotas commented on src/libraries/System.IO.MemoryMappedFiles/src/System.IO.MemoryMappedFiles.csproj:
```diff
@@ -132,32 +139,31 @@
   <ItemGroup>
     <Reference Include="System.IO.FileSystem" />
-    <Reference Include="System.Runtime" />
+    <Reference Include="System.Resources.ResourceManager" />
```
`System.Resources.ResourceManager` should not be needed

jkotas commented on src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.Unix.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace System.IO.MemoryMappedFiles
+{
+    public static class MemoryMappedFileAcl
```
Would it be better to let these implementations to be auto-generated in System.IO.FileSystem.AccessControl with a better PlatformNotSupportedException?

carlossanlop replied:
What's the command to auto generate the files?

safern replied:
I don't think we have a way to auto-generate only a list of APIs, we can only auto-generate the not supported assemblies using GenAPI at compile time for all the APIs defined in the ref assembly. @ericstj can correct me if I'm wrong.

jkotas commented:
I do not think it is a good trade-off to introduce all this forwarding, layering inversion between ref and src, etc. just to avoid introducing one extra public API to make this work cleanly.

carlossanlop commented:
> _I do not think it is a good trade-off to introduce all this forwarding, layering inversion between ref and src, etc. just to avoid introducing one extra public API to make this work cleanly._

The most recent changes would address the [first suggested option in the API Review](https://github.com/dotnet/runtime/issues/33206). I had other 2 suggestions, before applying the third one. Would you agree with one of the other two options? cc @stephentoub @bartonjs

jkotas commented:
Why do we need the API with that problematic `leaveOpen` argument at all?

stephentoub commented:
> Why do we need the API with that problematic leaveOpen argument at all?

@jkotas, is your question why do we need the overload that takes a `FileStream`?  From my perspective, _if_ we provide one that takes a `FileStream`, we'd need to either provide the `leaveOpen` argument, or not provide it and have the default behavior be `leaveOpen == false`, which is the default behavior in other APIs when such a `leaveOpen` isn't specified; from a consistency standpoint, it'd be concerning to me to have the behavior be equivalent to `leaveOpen == true`... and it's the `leaveOpen == false` behavior that's more challenging here.  If the question was about having the `FileStream`-based overload at all, I'll let @carlossanlop answer; I'm not familiar with the exact customer requests driving adding these APIs.

> I do not think it is a good trade-off to introduce all this forwarding, layering inversion between ref and src, etc. just to avoid introducing one extra public API to make this work cleanly.

Don't we already have the same layering in the equivalent APIs for System.IO.Pipes?  Seems desirable to employ the same solution in both places; if we don't want to go the expose-it-from-a-higher-level-ref-but-implement-it-in-the-main-library approach, we should ideally change it in Pipes, too.

> The most recent changes would address the first suggested option in the API Review. I had other 2 suggestions, before applying the third one. Would you agree with one of the other two options?

IIRC, the others were actually the preferred approach as discussed in the API review, e.g. unseal SafeMemoryMappedFileHandle, and have an internal derived type in the access control library that stores the FileStream and overrides ReleaseHandle to Dispose of it as well.  Then the MemoryMappedFile ctor would just take a SafeMemoryMappedFileHandle.

That said, looking at it again, it seems problematic to add any such constructor.  MemoryMappedFile currently doesn't have any constructors, instead having the various CreateXx methods.  Adding a new ctor, IMHO, is going to make the type harder to use, because it'll then have a constructor, albeit a very complicated one that no one should be using, and it'll make it less likely that folks will understand they need to use the factories.  We should probably follow-suit, and have a CreateFromHandle method, or something similarly named.

Taking a step back, how important are these new APIs at all?  Are we adding these just for completeness, or are there developers that have been asking for these for real needs?

carlossanlop commented:
> _Why do we need the API with that problematic `leaveOpen` argument at all?_

When the user calls `MemoryMappedFile.CreateFromFile` we need to root the passed `fileStream`.

If the user called `CreateFromFile` with `leaveOpen = false`, then the rooted `fileStream` needs to be disposed when the `MemoryMappedFile` is disposed.

If the user called `CreateFromFile` with `leaveOpen = true`, the rooted `fileStream` will not get disposed when the `MemoryMappedFile` is disposed, but we ensured it will remain alive because it was rooted.

jkotas commented:
Why can't we have just the API that leaves the FileStream alone (just gets the handle from it) and always make the user responsible for disposing the FileStream?

stephentoub commented:
> Why can't we have just the API that leaves the FileStream alone (just gets the handle from it) and always make the user responsible for disposing the FileStream?

Pretty much every other API in the framework that takes a Stream takes ownership of the stream by default, unless a leaveOpen parameter is used and set to true.

jkotas commented:
If we are taking ownership of the stream, can we close it as soon as we get the handle out of it and not wait until the memory mapping is closed?

stephentoub commented:
> If we are taking ownership of the stream, can we close it as soon as we get the handle out of it and not wait until the memory mapping is closed?

That would differ in behavior from the existing almost-identical overloads that take a FileStream (and leaveOpen) and just don't take the ACL-related args.  It'd be observably inconsistent to differ in that regard.  But whether it would cause anyone problems, I don't know.

jkotas commented:
> Don't we already have the same layering in the equivalent APIs for System.IO.Pipes?

Ok, I have not realized that. I agree that both should be on the same plan.

carlossanlop commented:
@stephentoub

> _IIRC, the others were actually the preferred approach as discussed in the API review, e.g. unseal SafeMemoryMappedFileHandle_

Those were options 2.1 and 2.2. These were the options in [Immo's approval comment](https://github.com/dotnet/runtime/issues/33206#issuecomment-610526663):

> We're fine with taking the handle parameter but we have concerns with the fileStream parameter because it seems plumbing, not part of a sensible API.
> Option 1: Move the implementation from the higher level assembly to the assembly containing MemoryMappedFile (while the ref assembly setup)
> Option 2.1 "Smuggle" the file stream via the SafeMemoryMappedFileHandle via a field
> Option 2.2 "Smuggle" the file stream via the SafeMemoryMappedFileHandle unsealing & deriving from SafeMemoryMappedFileHandle
> Option 3: do what's proposed

> _how important are these new APIs at all? Are we adding these just for completeness, or are there developers that have been asking for these for real needs?_

These ACL APIs are being added for completeness. @JeremyKuhne created API Proposals for all the types that had similar ACL needs. All others have been address, and this one for `MemoryMappedFile` had the lowest priority because there was no user request.

Should I attempt option 2.1 instead?

stephentoub commented:
> because there was no user request

If literally no one wants it, we shouldn't spend any more time on it ;-)

danmosemsft commented:
> If literally no one wants it, we shouldn't spend any more time on it ;-)

I'd potentially agree if we didn't support use of ACL's on MemoryMappedFiles, but in this case we are currently offering only an API that invites insecure use. We shouldn't need a customer request to add back API to make it possible to do the job securely. As it stands, a successful port to .NET Core of code using ACL's on MMF (apisof.net shows folks have run the apiport tool on such code) may necessitate changing the code to make it less secure.

danmosemsft commented:
Having said that -- if we think doing this is not going to help those customers be successful (eg because we're not happy with the API) -- then I certainly am not arguing to do the wrong thing 😄  I think it would be better to sit on the problem until we feel confident we can do something useful.

jkotas commented:
There are many ways to obtain file mapping handle: opening or creating it with special security permissions, or creating it using a specialized API (e.g. `CreateFileMappingNuma`), or get it by cloning out of process handle, or ... . The first thing we should be doing in situation like this is to build the simplest possible extension point that allows people to be successful: Add a constructor that takes `SafeMemoryMappedFileHandle` that covers all these cases. People may need to use P/Invoke to obtain the desired file mapping handle, but the rest will be handled by the framework.

Only once we have that we should be looking at the specialized construction paths, like the path that takes ACLs in this PR.

carlossanlop commented:
I will close the PR. We can come back to it if a customer requests it.

danmosemsft commented:
OK, thanks @carlossanlop
